File: Graph.cs
using System.Collections.Generic;

/// <summary>
/// Base class for all graphs
/// </summary>
public class Graph
{
	// Determines if the graph is two-way
    bool bidirectional;
	
	// Used to look up nodes based on IDs
    Dictionary<int, Node> nodeLookup = new Dictionary<int, Node>();

    protected class Node
    {
        public int ID { get; private set; }
        public List<Node> adjacent = new List<Node>();

        public Node(int id)
        {
            ID = id;
        }
    }

    public Graph()
    {
        bidirectional = true;
    }

    public Graph(bool bidirectional)
    {
        this.bidirectional = bidirectional;
    }

	// Returns a node from an id
	// If one doesn't exist, creates a new one
    protected Node GetNode(int id)
    {
        if (nodeLookup.ContainsKey(id))
            return nodeLookup[id];
        else
        {
            Node newNode = new Node(id);
            nodeLookup[id] = newNode;
            return newNode;
        }
    }

    public void AddEdge(int start, int destination)
    {
        Node s = GetNode(start);
        Node d = GetNode(destination);
		
        s.adjacent.Add(d);
        if (bidirectional)
            d.adjacent.Add(s);
    }
}

File: Graph.cpp
// A base Graph class which is included in unweighted algorithms

#include <unordered_map>
#include <set>

struct Node
{
    std::set<Node*> adjacent;	
    int ID;
};

class Graph
{
private:
	// Determines if the graph is directed
	// (edges are one-way)
    bool directed;
	
	// Keeps track of nodes based on IDs
    std::unordered_map<int, Node> nodeLookup;

protected:
	// Returns node from nodeLookup
	// Creates a new one if it doesn't exist
    Node* getNode(int id)
    {
        if (nodeLookup.find(id) != nodeLookup.end())
            return &nodeLookup[id];
        else
        {
            Node newNode;
            newNode.ID = id;
            nodeLookup[id] = newNode;
            return &nodeLookup[id];
        }
    }

public:
    Graph(bool direct) : directed(direct) {}

    void addEdge(int source, int destination)
    {
        Node* s = getNode(source);
        Node* d = getNode(destination);
		
        s->adjacent.insert(d);
        if (!directed)
            d->adjacent.insert(s);
    }
};
